File: Kalorie/Kalorie/Core/UseCases/FetchFavouriteFoodsUseCase.swift
//
//  FetchFavouriteFoodsUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 12.08.2026.
//

import Foundation

protocol FetchFavouriteFoodsUseCaseProtocol {
    func callAsFunction() async throws -> [FoodItemDomain]
}

struct FetchFavouriteFoodsUseCase: FetchFavouriteFoodsUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction() async throws -> [FoodItemDomain] {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let dtos: [FavouriteFoodDTO] = try await dataProvider.loadAsync(
            from: Constants.Firestore.favouriteFoods(userId: userId),
            orderBy: "favourited_at",
            descending: true,
            limit: 50
        )
        return dtos.map { $0.asDomain() }
    }
}

#if DEBUG
struct FetchFavouriteFoodsUseCaseFake: FetchFavouriteFoodsUseCaseProtocol {

    // MARK: - Properties

    var stubbedItems: [FoodItemDomain] = []
    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction() async throws -> [FoodItemDomain] {
        if shouldThrow { throw URLError(.unknown) }
        return stubbedItems
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/FetchFavouriteFoodsUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.networking.FavouriteFoodDTO
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.loadAsync
import antoni.kalorie.core.utils.Constants

interface FetchFavouriteFoodsUseCaseProtocol {
    suspend operator fun invoke(): List<FoodItemDomain>
}

class FetchFavouriteFoodsUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : FetchFavouriteFoodsUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(): List<FoodItemDomain> {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val dtos: List<FavouriteFoodDTO> = dataProvider.loadAsync(
            from = Constants.Firestore.favouriteFoods(userId),
            orderBy = "favourited_at",
            descending = true,
            limit = 50,
        )
        return dtos.map { it.asDomain() }
    }
}
